docs/howto/csrf-protection.txt

```diff
@@ -0,0 +1,321 @@
+.. _using-csrf:
+
+===================================
+How to use Django's CSRF protection
+===================================
+
+To take advantage of CSRF protection in your views, follow these steps:
+
+1. The CSRF middleware is activated by default in the :setting:`MIDDLEWARE`
+   setting. If you override that setting, remember that
+   ``'django.middleware.csrf.CsrfViewMiddleware'`` should come before any view
+   middleware that assume that CSRF attacks have been dealt with.
+
+   If you disabled it, which is not recommended, you can use
+   :func:`~django.views.decorators.csrf.csrf_protect` on particular views
+   you want to protect (see below).
+
+2. In any template that uses a POST form, use the :ttag:`csrf_token` tag inside
+   the ``<form>`` element if the form is for an internal URL, e.g.:
+
+   .. code-block:: html+django
+
+       <form action="" method="post">{% csrf_token %}
+
+   This should not be done for POST forms that target external URLs, since
+   that would cause the CSRF token to be leaked, leading to a vulnerability.
+
+3. In the corresponding view functions, ensure that
+   :class:`~django.template.RequestContext` is used to render the response so
+   that ``{% csrf_token %}`` will work properly. If you're using the
+   :func:`~django.shortcuts.render` function, generic views, or contrib apps,
+   you are covered already since these all use ``RequestContext``.
+
+.. _csrf-ajax:
+
+Using CSRF protection with AJAX
+===============================
+
+While the above method can be used for AJAX POST requests, it has some
+inconveniences: you have to remember to pass the CSRF token in as POST data with
```

Comment:
Not just that, but it is simply impossible if you are posting json.

Reply:
Why is this impossible in that case?

Reply:
Because this line: https://github.com/django/django/blob/d90936f41a2d7a3361e51fc49be033ba0f05f458/django/middleware/csrf.py#L289 will not work for JSON, you have to use the header then.